addons/account/models/account_move_line.py

```diff
@@ -1847,94 +1840,482 @@ def get_accounting_rate(vals):
+        return all_results
+
+    @api.model
+    def _optimize_reconciliation_plan(self, reconciliation_plan):
+        """ Decode the initial reconciliation plan passed as parameter and converted it into a list of tree depicting
+        the way the reconciliation should be done.
+        Also, this method is responsible sorting the amls and splitting them by currency.
+        Then, this method checks the parameter to ensure we are not going to perform any invalid reconciliation like
+        a cross-account/cross-company partial.
+
+        The split by currencies is made as follows.
+        Suppose account.move.line(1, 2) are expressed in currency1 and account.move.line(3, 4) are expressed
+        in currency2.
+        If the reconciliation plan is [account.move.line(1, 2, 3, 4)], the optimizer will convert it into:
+        [[account.move.line(1, 2), account.move.line(3, 4)]]
+
+        :param reconciliation_plan: A list of reconciliation to perform.
+        :return: A list of dictionaries containing:
+            * amls: A recordset.
+            * aml_ids: The recordset ids.
+            * nodes: A list of sub-nodes.
+        """
+
+        def process_amls(amls):
+            sorted_amls = amls.sorted(key=lambda aml: (
+                aml.date_maturity or aml.date,
+                aml.currency_id,
+                aml.amount_currency,
+                aml.balance,
+            ))
+            currencies = sorted_amls.currency_id
+
+            results = {
+                'amls': sorted_amls,
+                'aml_ids': set(sorted_amls.ids),
+            }
+
+            if len(currencies) != 1:
+                nodes = results['nodes'] = []
+                for currency in currencies:
+                    amls_in_currency = sorted_amls.filtered(lambda x: x.currency_id == currency)
+                    nodes.append({
+                        'amls': amls_in_currency,
+                        'aml_ids': set(amls_in_currency.ids),
+                    })
+            return results
+
+        def process_children(children):
+            node = {
+                'nodes': [],
+                'aml_ids': set(),
+            }
+            for child in children:
+                results = process_leaf(child)
+                if results:
+                    node['nodes'].append(results)
+                    node['aml_ids'].update(results['aml_ids'])
+            return node
+
+        def process_leaf(item):
+            if not item:
+                return
+
+            if isinstance(item, models.BaseModel):
+                # Group of amls to evaluate.
+                return process_amls(item)
+            else:
+                # Sub plan to evaluate.
+                return process_children(item)
+
+        plan_list = []
+        all_aml_ids = set()
+        for item in reconciliation_plan:
+            plan_node = process_leaf(item)
+            if not plan_node or not plan_node.get('amls'):
+                continue
```

Comment:
`not plan_node.get('amls')` wasn't there previously.
I'm not sure why it was added